umbim / wwan: bugfixes + support for non-dhcp devices

sch-m opened the pull request:
The commits in the PR will fix some problems (cleanup on error + pinstate validation) in the umbim proto handler and add the support for mbim capable wwan modules, which do not support ip autoconfiguration via dhcp (like the gemalto Cinterion ELS81).

sch-m committed:
umbim: add support for non-dhcp mode

There are mbim compatible wwan modules available which do not support
the dhcp autoconfiguration. (e.g. gemalto Cinterion ELS81)

This adds the possibility to get the configuration parameters from mbim.

Signed-off-by: Martin Schiller <ms@dev.tdt.de>

sch-m committed:
wwan: add gemalto Cinterion ELS61/ELS81 in MBIM mode to database

Signed-off-by: Martin Schiller <ms@dev.tdt.de>

sch-m committed:
wwan: add new flag dhcp to database

This flag adds the possibilty to tell the "real" proto handler if the
used device supports dhcp or not.

Signed-off-by: Martin Schiller <ms@dev.tdt.de>

sch-m committed:
wwan: bump PKG_RELEASE

bump PKG_RELEASE because of my former changes.

Signed-off-by: Martin Schiller <ms@dev.tdt.de>

adrianschmutzler commented:
@blogic @jow Can you have a look?

sch-m committed:
wwan: add Thales Cinterion MV31-W in MBIM mode to database

Signed-off-by: Martin Schiller <ms@dev.tdt.de>

Leo-PL commented:
@sch-m could you please rebase this? I'd like to do some testing with Sierra EM7455, and I'd be quite happy to see this PR merged.

sch-m commented:
> @sch-m could you please rebase this? I'd like to do some testing with Sierra EM7455, and I'd be quite happy to see this PR merged.

done.

Leo-PL commented:
Tested-By: Lech Perczak <lech.perczak@gmail.com>
Works fine with my Sierra EM7455 and Archer C7v2. 

I noticed one issue, probably unrelated to your PR, but possibly fixable here as well: Independent of IP type selection in config (v4 only, v6 only or v4v6), both IPv4 and IPv6 are always brought up. This is also true with static configuration, when only one of sub-interfaces actually has configuration provided from the PDN.

I observed this both on master branch, as well as in backport I attempted at 19.07.

Leo-PL commented on package/network/utils/umbim/files/lib/netifd/proto/mbim.sh:
```diff
@@ -121,27 +135,69 @@ _proto_mbim_setup() {
+		proto_send_update "$interface"
+
+		json_init
+		json_add_string name "${interface}_4"
```
While already touching this - as I mentioned in the comment, I would selectively enable sub-interfaces basing on IP type selection.
Otherwise - LGTM.

hyc commented:
Compared to #3317 this patch lacks setting the pdptype on the connect request, as well as conditional enabling of IPv4 or IPv6 interfaces. Should be simple enough for you to copy what I did there into this patch.

Leo-PL commented:
@hyc That's exactly what I noticed as well, and made implemented it here: https://github.com/Leo-PL/openwrt/commit/b9f747e46fd694e50349bb70ff04842ffe59cb7f - I need to test it myself, though.

I would like to see #3317 merged as well, IMO this implementation is cleaner, but it does need up splitting into smaller chunks, just for ease of maintenance - it's easier to revert a small patch than a big one, should something break.

Rebasing it over this PR, though, would probably leave your refactoring as the last remaining change, as all the minor fixes in both PRs are mostly the same.

hyc commented:
I've added my changes on top of this patch:
[0003-Also-don-t-use-DHCPv6-unless-IPv6-was-enabled.patch.txt](https://github.com/openwrt/openwrt/files/6835159/0003-Also-don-t-use-DHCPv6-unless-IPv6-was-enabled.patch.txt)
[0002-Check-for-IPv4-or-IPv6-support-before-configuring-re.patch.txt](https://github.com/openwrt/openwrt/files/6835160/0002-Check-for-IPv4-or-IPv6-support-before-configuring-re.patch.txt)
[0001-Add-pdptype-param-for-selecting-IPv4-IPv6-or-IPv4v6.patch.txt](https://github.com/openwrt/openwrt/files/6835161/0001-Add-pdptype-param-for-selecting-IPv4-IPv6-or-IPv4v6.patch.txt)

hyc commented:
I'm not convinced that keeping a dhcp flag in the modem database is useful, since it's the network carrier that determines how IP configuration is done.

Leo-PL commented:
I'm not so sure about that,
FWIW the DHCP server is optionally ran on the modem itself, be it for NCM, QMI or MBIM, given that PDN links are IP links tunneled directly inside PDCP - no way to send broadcast there.

Leo-PL commented:
> 
> 
> I'm not convinced that keeping a dhcp flag in the modem database is useful, since it's the network carrier that determines how IP configuration is done.

@hyc Also, according to MBIM specification, configuration through DHCP/DHCPv6 or SLAAC is also supported, if "IP Configuration information elements" are not provided via MBIM, but if they are present, they shall be preferred. But still, this is usually served by the modem.

@sch-m I actually fixed that comment I made myself, along fixing couple more bugs and improving DHCP/static configuration coexistence here: https://github.com/Leo-PL/openwrt/tree/mbim-fixes.

sch-m commented:
@hyc @Leo-PL 
So now we have some patches [here](https://github.com/TDT-AG/openwrt/pull/1) and [there](https://github.com/Leo-PL/openwrt/commits/mbim-fixes), but what exactly should we do with them now?

Do you want me to integrate/add them into/to this PR?

hyc commented:
If @Leo-PL prefers his version of the patches, go ahead and use his. I'm currently running with https://github.com/TDT-AG/openwrt/pull/1 but I don't much care which is finally used.

Leo-PL commented:
@sch-m I'm fine with either option as well - you may pull my patches in, or I can rebase my patches to current master - I have to do couple of builds as well anyway - and open a new PR. I think it may be easier if you cherry-pick my changes and test them as well.

Leo-PL commented:
@sch-m @hyc I updated my changes, to also keep DHCP enabled by default, because of two reasons:
- To not break existing users possibly using DHCP, which was previously the only available method,
- MBIM specification 1.0 mentions that IPv6 configuration is most frequently provided inband.

I also added check for "ipv6" flag as in #3317, but also checking if IPv6 is enabled in kernel, falling back to IPv4 PDP type if not. I also rebased them and tested against current master. I think it'll be easiest for maintainers if I open my PR with them. What do you think? I also kindly ask for testing those changes.

LuCI plugin will have to accomodate for the new defaults, both "dhcp" and "ipv6" flags are enabled by default.

hyc commented:
AFAICS there cannot have been any existing users, so I don't believe backward compatibility is a valid issue. Particularly for Luci there clearly were never any users before.

Leo-PL commented:
This is not only about backwards compatibility. Apparently, DHCP-supporting modems exist, and somebody was using MBIM protocol with OpenWrt, since this support was available in 19.07 release.

My point is to support most usecases without need for explicit configuration, and enabling both DHCP and IPv6 by default attains that.

MBIM specification [1] states that if IP configuration is not provided through MBIM_CID_IP_CONFIGURATION, then in-band IP configuration mechanisms shall be used:

> Sometimes, some or all of the IP configuration information is obtained as part of the procedure in which
the link-layer for data services is established. In such cases, the MBIM function acquires IP configuration
information from the carrier network on behalf of the host. This CID is used to transfer such IP
configuration information from an MBIM function to the host.
> It is also possible for some or all of the IP configuration information to be acquired by the host’s IP stack
interacting directly at or above the IP layer with the carrier network. In this case, the MBIM function
does not participate in the acquisition of IP configuration information other than enabling basic IP
packet service.
> Thus, the two alternatives are:
> - The MBIM function obtains IP configuration information on behalf of the host from the carrier
network (more common for IPv4).
> - The host’s IP stack obtains IP configuration information from the carrier network (more
common for IPv6).

Example of this is the fact, that Sierra EM7455 is able to support DHCP passthrough on its data connection - see AT!DHCPRELAYENABLE command in its AT command reference.

Given the next paragraph:
> It is also possible to obtain some IP configuration information via the MBIM function, and some other
via the host’s IP stack. In particular, some deployments of IPv6 may still rely on the MBIM function to
obtain the DNS information, while relying on the host’s IP stack to acquire the other elements of IP
configuration information.

I think we should also include MBIM-provided DNS servers in the list for DHCP sub-interfaces if they are present, but address information isn't. But let's wait with that until I get an IPv6-enabled SIM card for testing, probably in the next week, and get MBIM decently operational first.

[1] Mobile Broadband Interface Model v1.0 Errata -1 and Adopters Agreement  https://www.usb.org/document-library/mobile-broadband-interface-model-v10-errata-1-and-adopters-agreement

bobafetthotmail commented:
Did you already try moving this package to community  packages feed https://github.com/openwrt/packages and one of you take over as maintainer? After you did that you can easily fix and update it without waiting for core developer review like packages in this repository.
Other packages that were not maintained in core anymore nor included in default build of any device got moved in the past.

Leo-PL commented:
> 
> 
> Did you already try moving this package to community packages feed https://github.com/openwrt/packages and one of you take over as maintainer? After you did that you can easily fix and update it without waiting for core developer review like packages in this repository.
> Other packages that were not maintained in core anymore nor included in default build of any device got moved in the past.

Not yet. This could actually be an option, because there are no devices requiring `umbim` for their basic connectivity, but for this type of package - I really think it belongs in core.

hyc commented:
Who is responsible for / able to review this?

dwmw2 committed:
mediatek: mt7623: Disable power button reset for U7623-02 board

The Unielec U7623 doesn't have a physical power button; I think it's hard
wired so that it turns on automatically when power is applied (unlike the
Banana Pi R2 which is a pain).

So the 'reset on long press of power button' behaviour that we get when
we enable the PMIC keyboard driver is kind of unhelpful. Disable it.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>
Fixes: 0d3f3323a2 ("mediatek: mt7623: enable more hardware features")

hauke commented:
I merged #4405.
If this pull request is still needed please rebase it, otherwise please clone it.
